Review

hubduing opened the pull request:
Add basic logic.

UralBAV committed:
Add open form

UralBAV committed:
Change form article

UralBAV committed:
fix(изменение метода изменения размера текста): изменение записи размера текста в стейт

UralBAV committed:
feat(new tempstate): add tonew tempState

UralBAV committed:
feat(add css.): добавление css на кнопки, селекторы и т.д

olgalatkina commented on src/components/arrow-button/ArrowButton.tsx:
```diff
@@ -1,19 +1,34 @@
+	callBack: () => void;
+};
+export const ArrowButton: React.FC<OnClick> = ({ callBack }) => {
+	const arrowButtonRef = useRef<HTMLDivElement | null>(null);
```
**Надо исправить:**
В этом компоненте не понадобятся ни `useRef`, ни `useEffect`. 
Для условного добавления css-классов достаточно воспользоваться библиотекой `clsx` (уже есть в проекте). Это требование есть в чек-листе: https://ru.paste.pics/89320309e689ee231362c674cd5bd828

Возможно, в этой статье вы найдёте больше информации: https://habr.com/ru/articles/649381/.

olgalatkina commented on src/components/article-params-form/ArticleParamsForm.module.scss:
```diff
@@ -16,8 +16,9 @@
 	display: flex;
 	flex-shrink: 0;
 	flex-direction: column;
+	row-gap: 5vh;
```
**Надо исправить:**
По макету расстояние между элементами формы должно быть 50px. Динамическая величина не решает проблемы маленьких экранов, потому что минимальное расстояние получается 50px, а на больших форма слегка расползается и не понятна выгода от использования `vh`.

olgalatkina commented on src/index.tsx:
```diff
@@ -1,31 +1,67 @@
 const root = createRoot(domNode);
 
 const App = () => {
+	const [formState, setFormState] = useState(defaultArticleState); // Основной стейт.
```
**Надо исправить:**
Каждый компонент сам должен отвечать за хранение и обработку своего состояния. Использовать подъем состояния формы в родительский компонент не имеет смысла для нашего приложения. Состояние формы нужно описать в самом компонент `ArticleParamsForm`, а не передавать его из `App`, т.к. это состояние описывает именно `ArticleParamsForm`. В состоянии `App` должны хранится только данные о текущем отображении статьи.

olgalatkina commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,16 +1,108 @@
+			<aside
+				ref={asideRef}
+				className={clsx(styles.container, open && styles.container_open)}>
+				<form className={styles.form} onSubmit={apply}>
```
**Отлично:**
Отлично, что обработчик сохранения данных вызывается при отправке формы.

olgalatkina commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,16 +1,108 @@
+			<ArrowButton callBack={arrowButtonHandler} />
+			<aside
+				ref={asideRef}
+				className={clsx(styles.container, open && styles.container_open)}>
```
**Отлично:**
Замечательно, что для условного добавления css-классов используется библиотека `clsx`.

olgalatkina commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,16 +1,108 @@
+	const [open, setOpen] = useState(false);
+
+	// открытие закрытие боковой панели
+	const arrowButtonHandler = useCallback(() => {
```
**Надо исправить:**
Нет необходимости создавать дополнительные функции обработчики если они сами по себе ничего не делают, можно использовать просто функции сеттеры возвращаемые `useState`.
Тем более нет необходимости оборачивать в `useCallback` функции сеттеры, потому что ссылка на них является стабильной.

olgalatkina commented on src/index.tsx:
```diff
@@ -1,31 +1,67 @@
+	};
 	return (
 		<div
 			className={clsx(styles.main)}
```
**Можно лучше:**
Для одиночных селекторов не стоит использовать библиотеку `clsx` , лучше исправить этот момент здесь и в компоненте `Article`.

olgalatkina commented on src/index.tsx:
```diff
@@ -1,31 +1,67 @@
 const domNode = document.getElementById('root') as HTMLDivElement;
 const root = createRoot(domNode);
 
 const App = () => {
```
**Можно лучше:**
Обычно корневой `index.tsx` держат максимально компактным. Стоит вынести компонент `App` в отдельный файл в папке `components/app/App.tsx`.

olgalatkina reviewed (changes requested):
Здравствуйте.

**У вас получилась хорошая работа:**
- Полностью реализован функционал.
- Для закрытия сайдбара по клику вне сайдбара используется кастомный хук.
- Обработчик сохранения данных вызывается при отправке формы.

**Но есть несколько моментов, которые надо поправить:**
- [ ] Состояние формы нужно описать в самом компонент `ArticleParamsForm`.
- [ ] По макету расстояние между элементами формы 50px.
- [ ] Для условного добавления css-классов лучше использовать библиотеку `clsx`.
- [ ] Не стоит создавать дополнительные функции обработчики если они сами по себе ничего не делают.

**Правки можно лучше остаются на ваше усмотрение:**
- [ ] Для одиночных селекторов не стоит использовать библиотеку `clsx`.
- [ ] Стоит вынести компонент `App` в отдельный файл.

Успешного рефакторинга!

UralBAV committed:
refactor(articleparamsform, index, app, arrowbutton): перенос состояния. Вынесение App, clsx

Перенесено состояние из App в ArticleParamsForm. Исправлена работа с DOM, теперь используется clsx.
App теперь в scr/component/app

olgalatkina commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -21,47 +21,59 @@ import { Separator } from '../separator';
 	const asideRef = useRef<HTMLDivElement | null>(null);
 	const [open, setOpen] = useState(false);
+	const [formState, setFormState] = useState(defaultArticleState); // Основной стейт.
+	const [tempState, setTempState] = useState(formState); // Временный стейт, что бы опции сразу не применялись.
```
**Надо исправить:**
Этот стейт не нужен - он просто дублирует состояние формы, удаляем без сожаления. Используем стейт формы.

olgalatkina commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -21,47 +21,59 @@ import { Separator } from '../separator';
 }) => {
 	// хуки
 	const asideRef = useRef<HTMLDivElement | null>(null);
 	const [open, setOpen] = useState(false);
```
**Можно лучше:** 
Булевы значения принято называть с приставкой `is` , например `isFormOpen`.

olgalatkina commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -21,47 +21,59 @@ import { Separator } from '../separator';
 	const [open, setOpen] = useState(false);
+	const [formState, setFormState] = useState(defaultArticleState); // Основной стейт.
+	const [tempState, setTempState] = useState(formState); // Временный стейт, что бы опции сразу не применялись.
+	const apply = (e: FormEvent) => {
```
**Можно лучше:**
Для обработчиков событий принято в названии применять глагол `handle` или приставку `on`:
`on` -> `onSubmit`/`onReset`
`handle` -> `handleSubmit`/`handleReset`

olgalatkina commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -21,47 +21,59 @@ import { Separator } from '../separator';
+	};
+	const reset = () => {
+		setFormState(defaultArticleState);
+		setTempState(defaultArticleState);
```
**Надо исправить:**
При нажатии на кнопку СБРОСИТЬ должны сбрасываться состояние формы и состояние статьи до дефолтного.

olgalatkina commented on src/components/app/App.tsx:
```diff
@@ -0,0 +1,33 @@
+	// главное состояние
+	const [appState, setAppState] = useState(defaultArticleState);
+	// бумеранг
+	const updateState = (newState: ArticleStateType) => {
```
**Можно лучше:**
 Не стоит создавать дополнительные функции обработчики если они сами по себе ничего не делают. Передавайте в `ArticleParamsForm` функцию сеттер, у вас даже типизация не поменяется:
 ```
 <ArticleParamsForm updateState={setAppState} />
 ```

olgalatkina commented on src/index.tsx:
```diff
@@ -1,72 +1,10 @@
-
-import { Article } from './components/article/Article';
-import { ArticleParamsForm } from './components/article-params-form/ArticleParamsForm';
-import './styles/index.scss';
```
**Надо исправить:**
Эти стили надо вернуть. Иначе приложение выглядит не по макету и частично перестаёт работать.

olgalatkina commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -21,47 +21,59 @@ import { Separator } from '../separator';
+	const handleWidthAr = (selected: OptionType) => {
+		setTempState({ ...tempState, contentWidth: selected });
+	};
+	const handleFontSize = (selected: OptionType) => {
```
**Можно лучше:**
Попробуйте написать один обработчик вместо пяти почти одинаковых функций.

olgalatkina commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -21,47 +21,59 @@ import { Separator } from '../separator';
 		isOpen: open,
 		rootRef: asideRef,
-		onClose: arrowButtonHandler,
+		onClose: callBack,
```
**Надо исправить:**
Тут не переключение состояния, а установка значения в `false`:
```
onClose: () => setIsOpen(false)
```

olgalatkina commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -21,47 +21,59 @@ import { Separator } from '../separator';
 
-	// открытие закрытие боковой панели
-	const arrowButtonHandler = useCallback(() => {
+	const callBack = () => {
```
**Можно лучше:**
Не стоит создавать дополнительные функции обработчики если они сами по себе ничего не делают:
```
<ArrowButton isOpen={isOpen} onClick={() => setIsOpen(!isOpen)} />
```

olgalatkina commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -21,47 +21,59 @@ import { Separator } from '../separator';
 	return (
 		<>
-			<ArrowButton callBack={arrowButtonHandler} />
+			<ArrowButton open={open} callBack={callBack} />
```
**Можно лучше:**
`callBack` - не самое удачное название. В стартер-ките был дан тип, который был неплохой подсказкой:
```
export type OnClick = () => void;
```

olgalatkina reviewed (changes requested):
**Осталось поправить совсем немного:**
- [ ] Первым делом вернуть подключение стилей.
- [ ] Удалить дублирование состояния формы.
- [ ] Починить баг сброса состояния формы и статьи.

**Рекомендации можно лучше остаются на ваше усмотрение:**
1. Не стоит создавать дополнительные функции обработчики если они сами по себе ничего не делают.
2. Лучше поправить наименования.
3. Попробуйте написать один обработчик вместо пяти.

Успешного рефакторинга!

UralBAV committed:
refactor(app, index, articleparamsform): подключение стилей, фикс дублирование состояний, фикс бага

olgalatkina commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -32,89 +32,91 @@ export const ArticleParamsForm: React.FC<ArticleParamsFormType> = ({
-		setFormState(defaultArticleState);
-		setTempState(defaultArticleState);
+	const handleReset = () => {
+		updateState(defaultArticleState);
```
**Надо исправить:**
Не исправлено замечание прошлого ревью: _При нажатии на кнопку СБРОСИТЬ должны сбрасываться состояние формы и состояние статьи до дефолтного._
Этот пункт в ТЗ: https://yadi.sk/d/XVNFgek48heyhA

Вы сбрасываете до дефолтных настроек только состояние статьи. Нужно добавить сброс до дефолтных настроек у формы.

olgalatkina commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -32,89 +32,91 @@ export const ArticleParamsForm: React.FC<ArticleParamsFormType> = ({
 	};
+
+	// eslint-disable-next-line @typescript-eslint/no-unused-vars
+	// const handleFormState = (selected: OptionType, property: string) => {
```
**Можно лучше:**
Типизацию для `property` можно написать точнее: `keyof ArticleStateType`

Как вариант (не потому что лучше, просто вариант), этот общий обработчик можно написать так:
```
const handleOnChange = (type: keyof ArticleStateType) => {
	return (value: OptionType) => {
		setFormState((prevState) => ({ ...prevState, [type]: value }));
	};
};
```
его использование будет ещё компактнее:
```
<Select
  title={'шрифт'}
  onChange={handleOnChange('fontFamilyOption')}
  selected={state.fontFamilyOption}
  options={fontFamilyOptions}
/>
```

olgalatkina reviewed (changes requested):
Здравствуйте.

**Не исправлено замечание прошлого ревью:** При нажатии на кнопку СБРОСИТЬ должны сбрасываться состояние формы и состояние статьи до дефолтного.
Этот пункт в ТЗ: https://yadi.sk/d/XVNFgek48heyhA
Вы сбрасываете до дефолтных настроек только состояние статьи. Нужно добавить сброс до дефолтных настроек у формы.

Успешного рефакторинга!

UralBAV committed:
fix: default formState in Form

olgalatkina reviewed (approved):
Добрый день!

**Отличная работа! проект принимается**🎉

Успехов в дальнейшем обучении!
